runners/local-artifact-service-java/src/main/java/org/apache/beam/artifact/local/LocalFileSystemArtifactStagerService.java

```diff
@@ -0,0 +1,270 @@
+    }
+  }
+
+  File getArtifactFile(String artifactName) {
+    return new File(artifactsBase, artifactName);
+  }
+
+  private class CreateAndWriteFileObserver implements StreamObserver<PutArtifactRequest> {
+    private final StreamObserver<PutArtifactResponse> responseObserver;
+    private FileWritingObserver writer;
+
+    private CreateAndWriteFileObserver(StreamObserver<PutArtifactResponse> responseObserver) {
+      this.responseObserver = responseObserver;
+    }
+
+    @Override
+    public void onNext(PutArtifactRequest value) {
+      try {
+        if (writer == null) {
+          if (!value.getContentCase().equals(ContentCase.NAME)) {
+            throw Status.INVALID_ARGUMENT
+                .withDescription(
+                    String.format(
+                        "Expected the first %s to contain the Artifact Name, got %s",
+                        PutArtifactRequest.class.getSimpleName(), value.getContentCase()))
+                .asRuntimeException();
+          }
+          writer = createFile(value.getName());
+        } else {
+          writer.onNext(value);
+        }
+      } catch (StatusRuntimeException e) {
+        responseObserver.onError(e);
+      } catch (Exception e) {
+        responseObserver.onError(
+            Status.INTERNAL
+                .withCause(e)
+                .withDescription(Throwables.getStackTraceAsString(e))
+                .asRuntimeException());
+      }
+    }
+
+    private FileWritingObserver createFile(String artifactName) throws IOException {
+      File destination = getArtifactFile(artifactName);
+      if (!destination.createNewFile()) {
+        throw Status.ALREADY_EXISTS
+            .withDescription(String.format("Artifact with name %s already exists", artifactName))
+            .asRuntimeException();
+      }
+      return new FileWritingObserver(
+          destination, new FileOutputStream(destination), responseObserver);
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      if (writer != null) {
+        writer.onError(t);
+      } else {
+        responseObserver.onCompleted();
+      }
+    }
+
+    @Override
+    public void onCompleted() {
+      if (writer != null) {
+        writer.onCompleted();
+      } else {
+        responseObserver.onCompleted();
+      }
+    }
+  }
+
+  private static class FileWritingObserver implements StreamObserver<PutArtifactRequest> {
+    private final File destination;
+    private final OutputStream target;
+    private final StreamObserver<PutArtifactResponse> responseObserver;
+
+    private FileWritingObserver(
+        File destination,
+        OutputStream target,
+        StreamObserver<PutArtifactResponse> responseObserver) {
+      this.destination = destination;
+      this.target = target;
+      this.responseObserver = responseObserver;
+    }
+
+    @Override
+    public void onNext(PutArtifactRequest value) {
+      try {
```

Comment:
Random style thing - depending on the length of the monster `try { ... }` blocks, can be nice to separate into `onNextOrThrow(...) throws ...` so the `onNext` method is always a one-liner handling all the exceptions, and the main logic is more straight-line. This particular block is tiny so it is fine, but I can see it becoming relevant hereabouts.

Reply:
For the `commitManifest` call, which does much more work, done. For the rest, I'm not hugely concerned.